### ISSUE TITLE
[UX] Language switcher block: allow some basic customization via the block configuration form

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4924]
This was brought up as a forum question: https://forum.backdropcms.org/forum/language-switcher-how-change-looks-looking-short-tutorial-tips

> My new Backdrop website will be bilingual (Dutch/English). The default language switch block is huge and doesn't look nice:
>
> ![image](https://user-images.githubusercontent.com/2423362/106248794-29822980-621a-11eb-9c87-9407f50faad0.png)
>
> It is not so obvious to me how to tweak the appearance. I guess I have to use the configuration dialogue box for the Language switcher
> ...
> How do I get a simple NL / EN switch like this one:
>
> ![image](https://user-images.githubusercontent.com/2423362/106248889-4f0f3300-621a-11eb-86ce-430a0bc497d9.png)
>
> I guess the answer will be that I need to add some CSS style definitions, but how?

I would like to propose adding a couple of checkboxes to the block, to allow for:

- displaying language links as bullet points, or as a delimited list (using a pipe `|`)
- rendering the full name of the languages, or the respective language codes

**Comment by klonos** [comment]
I've got most of this working on my local. Here's where I'm currently at:

- This is what the configuration form looks like:

  ![image](https://user-images.githubusercontent.com/2423362/106461356-52662100-649d-11eb-8798-38d792c27814.png)

- There's some "clean up" before saving the block settings into config, so that they are saved as:

  ```json
  ...
  "block_settings": {
      "locale_switcher": {
          "language_list_display": "inline",
          "language_link_label_display": "code",
          "custom_delimiter": "/"
      }
  },
  ...
  ```

  ...instead of:

  ```json
  ...
  "block_settings": {
      "locale_switcher": {
          "language_list_display": "inline",
          "language_list_inline_delimiter": {
              "custom_delimiter": 1,
              "delimiter": "/"
          },
          "language_link_label_display": "code"
      }
  },
  ...
  ```

- Multiple instances of language switcher blocks can be added, and they may be configured differently/independently of each other:

  <img width="1181" alt="Screen Shot 2021-02-01 at 2 56 54 pm" src="https://user-images.githubusercontent.com/2423362/106461936-1b443f80-649e-11eb-9d0d-dae5e62ee0eb.png">

And now the blockers ...or at least things that are troubling me, and for which I'd like suggestions/opinions (+ some issues that should be follow-ups):

- The language switcher block uses `theme_links()` ([`theme('links__locale_block', $variables)` to be exact](https://github.com/backdrop/backdrop/blob/1.x/core/modules/locale/locale.module#L870)) ...this means that:
  - **Switching language links style**: can easily be achieved, simply by conditionally adding the `links` and `inline` classes to the generated list, which takes advantage of existing core CSS, and "magically" switches to inline instead of vertical bullet points 👍 ...although, can we really rely on that "magic" CSS to work on all/most themes? ...Bartik for example does render the links inline, but no space/delimiters between them 🤷 :
    <img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106469985-aa565500-64a8-11eb-8aa1-3a21523d0b25.png">
    ...or does it fall on the theme creator to ensure proper support for this? In other words, should we just make sure that the themes that ship with core work as expected, and then let contrib/custom themes figure things out?
  - **Specifying a custom delimiter** (when language links are configured to be rendered inline): Back in D4.7 (yes, D4!! 🙂 ) `theme_links()` [accepted a `$delimiter` parameter](https://git.drupalcode.org/project/drupal/-/blob/4.7.x/includes/theme.inc#L494), which defaulted to `' | '`. Unfortunately, although [theme_links should return list of links](https://www.drupal.org/project/drupal/issues/65151) did well in replacing the plain-text output with links, it also removed that `$delimiter` parameter, and over the course of time the delimiter was replaced with a "hard-coded" CSS `content: ',';`, which is added as `::after` in each `<li>` if they are being rendered inline (see: https://github.com/backdrop/backdrop/blob/1.x/core/themes/basis/css/component/small-text-components.css#L179) ...anyway, questions:
    - should we re-introduce this `$delimiter` parameter in the `theme_links()` function, as an additional option in the `$variables` array it accepts?
    - if we are extending the `theme_links()` function, should we also include an option to render links inline? (so that we can take advantage of this in lists other than the language switcher block)
    - speaking of changes, how do people feel re the bigger issue of the fact that `theme_item_list()` and `theme_links()` basically do the same thing (more or less) - relevant issue in Drupal-land: https://www.drupal.org/project/drupal/issues/2032645
- Should I create a new theme function specifically for the language switcher block? (and then handle all the vertical/inline, delimiter etc. in that theme function) ...that doesn't feel right to me, as the content of the language switcher block is basically a list of links, and that's what we have `theme_links()` and `theme_list_items()` for 🤔 
- If the answer to the previous point is "yes, new theme function", then what about existing sites that may already be using the `links__locale_block` theme suggestion to customize their language switchers? ...my idea was to replace the call to `theme('links__locale_block', $variables)` with a `theme('locale_block', $variables)`, but somehow still have `links__locale_block` theme suggestion take priority. I have tried adding `$variables['theme_hook_suggestions'][] = 'links__locale_block';` in `locale_preprocess_block()`, but that didn't seem to work 😞 ...I'll experiment some more.

I'll try to wrap most things up, and then file a PR, so that people can play with the block configuration form and provide feedback re terms/strings. As a lame themer myself 😅, I would welcome any feedback/ideas from avid themers of our community.

**Comment by klonos** [comment]
PS: I know that we could simply leave it up to themers to override the delimiter via CSS, but the point of this feature request is to allow customizing the look/style of the language switcher block via the admin UI, with as less CSS fiddling as possible.

**Comment by indigoxela** [comment]
Looking forward to test this. I can't provide anything to the decision task, though (what to add to core, what to leave to themers).

Let the language switcher block look less wretched OOTB makes sense to me. Why providing something, if we already know, that nobody can use it as-is. :wink:

**Comment by klonos** [comment]
Sorry for taking this long: https://github.com/backdrop/backdrop/pull/3524

What works:
- If you try to add a language block while only one language is enabled, you get a warning, with a useful link to the Languages page, to add/enable more languages for the site. (can split out to a separate issue if people find it scope creep)
- You can configure any language block to have the links show the full language name or the short language code.
- You can configure any language block to be vertical or horizontal
- If horizontal, you can optionally customize the delimiter character
- Settings are saved in config in a simplified form (no array nesting - not saving the delimiter checkbox value).
- The delimiter is added via a `data-*` attribute in each `li`, then `content: attr(data-list-delimiter);` is used in the CSS. (allows customization via the admin UI + can still be overridden with `content: ''` or whatever).
- `theme_links()` now accepts a new `'li_attributes'` key in its `$variables` array, to allow the `<li>`s in the generated `<ul>` to be more flexibly customized (can now add classes/attributes etc.) - this one will most likely need to be split out to a separate issue, but it was required in order to pass "dynamic" custom delimiters to the CSS for the list of languages (so that the delimiter can be tweaked via the admin UI). Happy to hear if anyone has an alternative that doesn't need all that. **Note**: the comma delimiter is still the default for anything that gets themed via `theme_links()` + gets the `.links` and `.inline` "magic" classes (that's [specific to Basis](https://github.com/backdrop/backdrop/blob/1.x/core/themes/basis/css/component/small-text-components.css#L179) though). So existing Basis-based themes won't break.

What doesn't work:
- There's still some craft here and there (a few commended-out lines).
- I somehow managed to break the block edit via the contextual links 😅 (goes to the correct layout "Manage blocks" page 👍 ...does not automatically open the block form 👎 ...still redirects to frontend as expected 👍 ). Looking into that regression.
- I will most likely need to fix tests (maybe also add new ones).

What I need feedback/thoughts on for:
- Still need to figure out how to best handle theme suggestions for the previously used `theme('links__locale_block')`, in order to ensure that we don't break existing sites with already-customized language switcher blocks.
- Need to figure out what color to use for the language delimiters (when links are rendered inline). For now, I've left it to `color: gray;` which works great with the Basis header (that otherwise uses a white shade for both links and plain text). Still a few remaining related things to figure out, like:
  - link underline on hover for language links etc. (the main nav menu items have a different background color on-hover 👍 ...the account menu links have underline 👍 ...the site name and language links don't have any effect on-hove OOTB (that's with Basis).
  - how things work in other core themes?
  - what about contrib themes? (would be nice to have some sane OOTB defaults)
- Need feedback re newly-introduced user-facing text (field labels, help text, warning messages etc.).
- Need feedback on newly-introduced function names, config properties, variable names etc.

That sums it up; I hope I didn't miss anything.

**Comment by indigoxela** [comment]
@klonos Many thanks for working on this issue!

Your PR causes multiple test failures, though. Things like `Undefined index: locale_switcher` should get fixed, before we can start testing. Back to "needs work".

**Comment by klonos** [comment]
Thanks for having a look @indigoxela 🙏 

Yes re tests. See how I said:

> - I will most likely need to fix tests (maybe also add new ones).

What I'm looking for now is feedback (and a consensus) on code implementation + also all the points I've listed in my previous comment + anything I've missed.

Test failures aside, and minus the regression with the context links, the actual feature (being able to configure the language switcher blocks to be horizontal/vertical etc.) works. So please test the PR sandbox, and let me know re UI strings, general feel of the workflow/UX, and things like color for delimiters etc.

Providing me with the above feedback will save me time from multiple back and forths, so thanks in advance for your time to review/test (even if not thorough at first pass).

**Comment by indigoxela** [comment]
> What I'm looking for now is feedback (and a consensus) on code implementation ... Providing me with the above feedback will save me time from multiple back and forths...

I see, agreed - consensus first. :+1:

**Comment by indigoxela** [comment]
I did some UX testing (incomplete), and I really like it! :+1: 

Two things might be worth considering:

1) It's allowed to set a *single* character as a delimiter for the horizontal display, but the used css actually only makes sense with pipes. A comma, for instance, looks awful. And maybe the restriction to one and only one character is too strict?

2) The color of that delimiter should be left to the theme(s), locale.css shouldn't provide padding or color IMO.

_Off topic:_ the old story of attribute `xml:lang` vs. `lang` vs. `hreflang` - is this something for a new issue - or is there already an existing one? (TLDR: locale uses `xml:lang` for the links, which is wrong.)

**Comment by BWPanda** [comment]
I might be a bit late to the party, but here's a contrib module I wrote for the one-and-only multi-lingual site I've worked on: https://github.com/backdrop-contrib/language_switcher_dropdown Maybe it (or its functionality) can be merged into this issue...?
